### PR TITLE
zero alloc compiler hooks

### PR DESCRIPTION
**Pull request by gretay-js**
~On top of https://github.com/ocaml-flambda/flambda-backend/pull/1407 starting from [Add Value.get_witnesses](https://github.com/ocaml-flambda/flambda-backend/commit/7ad012ec6122780d67f94f70f94e857dcd286cfb).~ 
Based on PR https://github.com/ocaml-flambda/flambda-backend/pull/858.

### CODE REVIEW

**Comment by gretay-js** [comment]
@poechsel 

Is there a particular reason that all calls to Compiler_hooks.execute in the backend are in Asmgen? If there is no strong reason, I'd prefer to put a hook to it in `Checkmach` to be called per function, rather than at the end of the analysis of the entire compilation unit. 

The main motivation is to reduce the memory footprint of this pass on the build. We are currently keeping all the details about all allocations in all functions until the end of the compilation unit. It's probably okay for most code but there may be pathological cases that would be nice to void.
Functions that remain unresolved until the end of the compilation unit analysis will have to hold on to their witnesses anyway.

Another minor reason is to get rid of the iter_witnesses type and just use Witneses.components  as an arguement for `Compiler_hooks.Show_allocations`.

**Comment by poechsel** [comment]
No strong reason, most likely that it was the less invasive way to add them.

Happy to have the hook called once per function if you think it's better!

**Comment by gretay-js** [comment]
> No strong reason, most likely that it was the less invasive way to add them.
> 
> Happy to have the hook called once per function if you think it's better!

It turns out to be a bit tricky (because of functions that remain unresolved untill the end of the module), so it's probably not worth it.

**Comment by poechsel** [comment]
I've tested the hooks in a separate project, everything seems to work as needed

File: driver/compiler_hooks.ml [hunk 1]
```diff
@@ -125,6 +128,9 @@ let register : type a. a pass -> (a -> unit) -> unit =
   | Cfg -> hooks.cfg <- f :: hooks.cfg
   | Cmm -> hooks.cmm <- f :: hooks.cmm
   | Inlining_tree -> hooks.inlining_tree <- f :: hooks.inlining_tree
+  | Check_allocations ->
+    Flambda_backend_flags.checkmach_details_cutoff := -1;
```
Comment (xclerc) [inline driver/compiler_hooks.ml:132]: I am a bit uneasy with the modification
of flag values here; could we rather use
`Misc.protect_refs` in `execute` below?

**Review by xclerc — APPROVED** [review]

**Review by mshinwell — APPROVED** [review]

**Pull request merged**